shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -27,6 +28,47 @@
 
 NSNotificationName const FlutterViewControllerWillDealloc = @"FlutterViewControllerWillDealloc";
 
+namespace {
+/// Provides a mechanism to pause a TaskRunner.
+class TaskRunnerPause {
+ public:
+  /// Create a `TaskRunnerPause`.
+  /// @param[in]  task_runner    The `TaskRunner` we will pause.
+  TaskRunnerPause(fml::RefPtr<fml::TaskRunner> task_runner) : task_runner_(task_runner) {}
+
+  /// Synchronously pauses the TaskRunnner.
+  void Pause() {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      is_paused_ = true;
+    }
+    fml::AutoResetWaitableEvent latch;
+    task_runner_->PostTask([&] {
+      latch.Signal();
+      std::unique_lock<std::mutex> lock(mutex_);
+      condition_variable_.wait(lock, [&] { return !is_paused_; });
+    });
+    latch.Wait();
```

Comment:
This seems like it could have two problems:

1) We unfortunately land here at some point where someone else is also locked waiting to get to the front of the io queue, causing a dealock. I _think_ that only happens during shell initialization, in which state we shouldn't be able to get here, but I'm not sure.
2) We take so long to finish getting through the rest of the IO queue that by the time we've paused it it's already managed to do work in the background.

Reply:
> 1. We unfortunately land here at some point where someone else is also locked waiting to get to the front of the io queue, causing a dealock. I _think_ that only happens during shell initialization, in which state we shouldn't be able to get here, but I'm not sure

I think you are referring to this: https://github.com/gaaclarke/engine/blob/c00f53d4b0e120740835552172a904fe987dd336/shell/common/shell.cc#L618:L618

That is why we have to call `Resume` before we call `surfaceUpdated:`.

> 2\. We take so long to finish getting through the rest of the IO queue that by the time we've paused it it's already managed to do work in the background.

I think @iskakaushik created a mechanism to post a task to the front of a message queue.  I can look into that.  The only other thing we could do is create a mechanism to cancel a decode job and restart it later.  That is a huge dangerous change.

Reply:
I talked with @iskakaushik.  We don't have a mechanism to post something to the front of a queue, he added that to Pipeline.

This PR turns a guaranteed crash to a crash if you have too many images to decompress at the time where the previous crash would happen.  So, this is is definitely an improvement and possibly good enough for 99.99% of practical usage.

Reply:
This kind of logic makes it much harder to reason about an already complicated system though.  It's currently susceptible to deadlocks, and that will likely only increase as time goes on.  

Would it help anything if we inverted things here?  Instead of trying to control the IO queue, we tell anyone trying to do OpenGL/graphics calls they have to check before they continue?